cmd/zoekt-merge-index/main.go

```diff
@@ -32,21 +33,119 @@ func merge(dstDir string, names []string) error {
 	return err
 }
 
-func main() {
-	paths := os.Args[1:]
+func mergeCmd(paths []string) error {
 	if paths[0] == "-" {
 		paths = []string{}
 		scanner := bufio.NewScanner(os.Stdin)
 		for scanner.Scan() {
 			paths = append(paths, strings.TrimSpace(scanner.Text()))
 		}
 		if err := scanner.Err(); err != nil {
-			log.Fatal(err)
+			return err
 		}
 		log.Printf("merging %d paths from stdin", len(paths))
 	}
 	err := merge(filepath.Dir(paths[0]), paths)
 	if err != nil {
-		log.Fatal(err)
+		return err
+	}
+	return nil
+}
+
+// explode splits a shard into indiviual shards and places them in dstDir.
+// If it returns without error, the input shard was deleted and the first
+// result contains the list of all new shards.
+//
+// explode cleans up tmp files created in the process on a best effort basis.
+func explode(dstDir string, inputShard string) error {
+	f, err := os.Open(inputShard)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	indexFile, err := zoekt.NewIndexFile(f)
+	if err != nil {
+		return err
+	}
+	defer indexFile.Close()
+
+	exploded, err := zoekt.Explode(dstDir, indexFile)
+	defer func() {
+		// best effort removal of tmp files. If os.Remove failes, indexserver will delete
+		// the leftover tmp files during the next cleanup.
+		for tmpFn := range exploded {
+			os.Remove(tmpFn)
+		}
+	}()
+	if err != nil {
+		return fmt.Errorf("exloded failed: %w", err)
+	}
+	var fns []string
+	for tmpFn, dstFn := range exploded {
+		err = os.Rename(tmpFn, dstFn)
+		if err != nil {
+			// clean up the shards we already renamed to avoid duplicate results.
+			for _, fn := range fns {
+				os.Remove(fn)
+			}
+			return fmt.Errorf("explode: rename failed: %w", err)
+		}
+		fns = append(fns, dstFn)
+	}
+
+	// Don't remove the input shard if its name matches one of the destination
+	// shards. This can happen, for example, if the input shard is a simple shard.
+	for _, dstFn := range exploded {
+		if dstFn == inputShard {
```

Comment:
did you test? Is there any chance that they don't match, ie one is a relative path and the other is absolute.

Reply:
yes, I tested it locally. Exploding simple shards updates the mod time of the file but leaves the shard in place. Exploding a  compound shard removes the compound shard, leaving only simple shards behind.